underfs/oss/src/main/java/alluxio/underfs/oss/OSSActiveSyncProvider.java

```diff
@@ -0,0 +1,356 @@
+    mExecutorService = Executors.newScheduledThreadPool(
+        ufsConf.getInt(PropertyKey.MASTER_UFS_ACTIVE_SYNC_THREAD_POOL_SIZE),
+        ThreadFactoryUtils.build("OSSActiveSyncProvider-%d", false));
+    mPollingThread = null;
+    mUfsUriList = new CopyOnWriteArrayList<>();
+    mEventMissed = false;
+    mTxIdMap = new ConcurrentHashMap<>();
+    mCurrentTxId = SyncInfo.INVALID_TXID;
+    mActiveUfsSyncMaxActivity = ufsConf.getInt(PropertyKey.MASTER_UFS_ACTIVE_SYNC_MAX_ACTIVITIES);
+    mActiveUfsSyncMaxAge = ufsConf.getInt(PropertyKey.MASTER_UFS_ACTIVE_SYNC_MAX_AGE);
+    mActiveUfsSyncInterval = ufsConf.getMs(PropertyKey.MASTER_UFS_ACTIVE_SYNC_INTERVAL);
+    mClient = ossClient;
+    mUfsUriLastModified = new ConcurrentHashMap<>();
+  }
+
+  /**
+   * Start the accounting for the next window of events.
+   * <p>
+   * This includes increasing the age of the unsynced syncpoints and decrease the activity of
+   * unsynced syncpoints.
+   */
+  private void initNextWindow() {
+    for (Map.Entry<String, Integer> activity : mActivity.entrySet()) {
+      String syncPoint = activity.getKey();
+      mActivity.put(syncPoint, activity.getValue() / 10);
+      mAge.put(syncPoint, mAge.get(syncPoint) + 1);
+    }
+  }
+
+  private void recordFileChanged(String syncPoint, String filePath, long txId) {
+    AlluxioURI syncPointUri = new AlluxioURI(syncPoint);
+
+    mChangedFiles.computeIfAbsent(syncPoint, (key) -> {
+      mActivity.put(syncPoint, 0);
+      mAge.put(syncPoint, 0);
+      mTxIdMap.put(syncPoint, txId);
+      return new ConcurrentHashSet<>();
+    });
+    try (LockResource r = new LockResource(mWriteLock)) {
+      mChangedFiles.get(syncPoint).add(
+          new AlluxioURI(syncPointUri.getRootPath() + filePath));
+      mActivity.put(syncPoint, mActivity.get(syncPoint) + 1);
+    }
+  }
+
+  private void syncSyncPoint(String syncPoint) {
+    mChangedFiles.remove(syncPoint);
+    mActivity.remove(syncPoint);
+    mAge.remove(syncPoint);
+    mTxIdMap.remove(syncPoint);
+  }
+
+  private void updateChangedFiles() {
+    for (AlluxioURI ufsUri : mUfsUriList) {
+      List<AlluxioURI> changedFiles = getChangedFiles(ufsUri);
+      for (AlluxioURI changedFile : changedFiles) {
+        recordFileChanged(ufsUri.toString(), changedFile.toString(), mCurrentTxId);
+      }
+    }
+  }
+
+  private List<AlluxioURI> getChangedFiles(AlluxioURI ufsUri) {
+    try {
+      if (isDirectory(ufsUri)) {
+        List<AlluxioURI> filesUri = new ArrayList<>();
+        String bucketName = UnderFileSystemUtils.getBucketName(ufsUri);
+        String prefix = cutSlash(ufsUri.getPath()) + "/";
+        ObjectListing objectListing = mClient.listObjects(bucketName, prefix);
+        List<OSSObjectSummary> sums = objectListing.getObjectSummaries();
+        for (OSSObjectSummary s : sums) {
+          if (isChanged(ufsUri, s.getLastModified().getTime())) {
+            updateUriLastModified(ufsUri, s.getLastModified().getTime());
```

Comment:
if there are several new files in the directory, maybe the better way is to get the latest timestamp and then update it.

Reply:
Instead of updating the last modified timestamp of the sync point, here we update the last modified timestamp of the files which are under the sync point.